ml-proto/spec/ast.ml

```diff
@@ -2,148 +2,67 @@
-type literal = value Source.phrase
+type target = target' Source.phrase
+and target' = Case of var | Case_br of var
 
 type expr = expr' Source.phrase
 and expr' =
-  | Nop                                     (* do nothing *)
-  | Block of expr list                      (* execute in sequence *)
-  | If of expr * expr * expr                (* conditional *)
-  | Loop of expr                            (* infinite loop *)
-  | Label of expr                           (* labelled expression *)
-  | Break of var * expr option              (* break to n-th surrounding label *)
-  | Switch of expr * var list * var * expr list   (* table switch *)
-  | Call of var * expr list                 (* call function *)
-  | CallImport of var * expr list           (* call imported function *)
-  | CallIndirect of var * expr * expr list  (* call function through table *)
-  | GetLocal of var                         (* read local variable *)
-  | SetLocal of var * expr                  (* write local variable *)
-  | Load of memop * expr                    (* read memory at address *)
-  | Store of memop * expr * expr            (* write memory at address *)
-  | LoadExtend of extop * expr              (* read memory at address and extend *)
-  | StoreWrap of wrapop * expr * expr       (* wrap and write to memory at address *)
-  | Const of literal                        (* constant *)
-  | Unary of unop * expr                    (* unary arithmetic operator *)
-  | Binary of binop * expr * expr           (* binary arithmetic operator *)
-  | Select of selop * expr * expr * expr    (* branchless conditional *)
-  | Compare of relop * expr * expr          (* arithmetic comparison *)
-  | Convert of cvt * expr                   (* conversion *)
-  | Unreachable                                    (* trap *)
-  | Host of hostop * expr list              (* host interaction *)
-
-
-(* Functions and Modules *)
-
-type memory = memory' Source.phrase
-and memory' =
-{
-  initial : Memory.size;
-  max : Memory.size;
-  segments : segment list;
-}
-and segment = Memory.segment Source.phrase
+  | Nop
+  | Block of labeling * expr list
+  | If of expr * expr
+  | If_else of expr * expr * expr
+  | Br_if of expr * var
+  | Loop of labeling * labeling * expr list
+  | Label of expr
+  | Br of var * expr option
+  | Return of var * expr option
+  | Tableswitch of labeling * expr * target list * target * expr list list
+  | Call of var * expr list
+  | Call_import of var * expr list
+  | Call_indirect of var * expr * expr list
+  | Get_local of var
+  | Set_local of var * expr
+  | Load of Kernel.memop * expr
+  | Store of Kernel.memop * expr * expr
+  | Load_extend of Kernel.extop * expr
+  | Store_wrap of Kernel.wrapop * expr * expr
+  | Const of Kernel.literal
+  | Unary of Kernel.unop * expr
+  | Binary of Kernel.binop * expr * expr
+  | Select of Kernel.selop * expr * expr * expr
+  | Compare of Kernel.relop * expr * expr
+  | Convert of Kernel.cvt * expr
```

Comment:
Would it be possible to move memop, literal, unop, binop, etc. out of Kernel and into Ast, and have Kernel import then from Ast instead? Since the Ast module is the thing that has the 1-1 correspondence with the design repo, it would be nice to have all the fields described in the design repo here, in one place.